gitlab-runner/src/client.rs

```diff
@@ -275,20 +282,72 @@ pub(crate) struct Client {
     client: reqwest::Client,
     url: Url,
     token: String,
+    system_id: String,
 }
 
 impl Client {
     pub fn new(url: Url, token: String) -> Self {
+        let system_id = Self::generate_system_id();
         Self {
             client: reqwest::Client::new(),
             url,
             token,
+            system_id,
+        }
+    }
+
+    fn generate_system_id_from_machine_id() -> Option<String> {
+        // Ideally this would be async for consistency, but that's for the next API bump really. In
+        // practise the client will be created at the start of the runner and the amount read is
+        // really tiny so blocking is not a real issue.
+        let mut f = match File::open("/etc/machine-id") {
+            Ok(f) => f,
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => return None,
+            Err(e) => {
+                warn!("Failed to open machine-id: {e}");
+                return None;
+            }
+        };
+
+        let mut id = [0u8; 32];
+        match f.read(&mut id) {
```

Comment:
could this be `read_exact` instead to avoid needing to check the length? Then this would only need to check for errors.